File: src/calibration/matrices.py
"""Load and save intrinsic calibration matrices."""

from __future__ import annotations

from pathlib import Path

import numpy as np

from src.config import load_config, load_section, resolve_config_path


def load_calibration_matrices(
    config_path: str | Path | None = None,
) -> tuple[np.ndarray, np.ndarray] | None:
    cal_cfg = load_section("calibration", config_path)
    if not cal_cfg.get("enabled", False):
        return None

    camera_matrix = np.array(cal_cfg["camera_matrix"], dtype=np.float64)
    dist_coeffs = np.array(cal_cfg["dist_coeffs"], dtype=np.float64).reshape(-1, 1)
    return camera_matrix, dist_coeffs


def save_calibration_matrices(
    camera_matrix: np.ndarray,
    dist_coeffs: np.ndarray,
    reprojection_error: float,
    image_size: tuple[int, int],
    config_path: str | Path | None = None,
) -> Path:
    path = resolve_config_path(config_path)
    config = load_config(path)

    config.setdefault("calibration", {})
    config["calibration"]["enabled"] = True
    config["calibration"]["camera_matrix"] = camera_matrix.tolist()
    config["calibration"]["dist_coeffs"] = dist_coeffs.reshape(-1).tolist()
    config["calibration"]["reprojection_error"] = reprojection_error
    config["calibration"]["image_width"] = image_size[0]
    config["calibration"]["image_height"] = image_size[1]

    from src.config import save_config

    return save_config(config, path)

File: src/calibration/matrices.cpp
#include "chess_robot/calibration/matrices.hpp"

#include "chess_robot/config.hpp"

namespace chess_robot {

std::optional<CalibrationMatrices> loadCalibrationMatrices(
    const std::optional<fs::path>& config_path) {
  YAML::Node cal = loadSection("calibration", config_path);
  if (!cal["enabled"] || !cal["enabled"].as<bool>()) {
    return std::nullopt;
  }
  if (!cal["camera_matrix"] || !cal["dist_coeffs"]) {
    return std::nullopt;
  }

  cv::Mat camera_matrix(3, 3, CV_64F);
  for (int r = 0; r < 3; ++r) {
    for (int c = 0; c < 3; ++c) {
      camera_matrix.at<double>(r, c) = cal["camera_matrix"][r][c].as<double>();
    }
  }

  const YAML::Node& dist_node = cal["dist_coeffs"];
  cv::Mat dist_coeffs(static_cast<int>(dist_node.size()), 1, CV_64F);
  for (std::size_t i = 0; i < dist_node.size(); ++i) {
    dist_coeffs.at<double>(static_cast<int>(i), 0) = dist_node[i].as<double>();
  }

  cv::Size image_size;
  if (cal["image_width"] && cal["image_height"]) {
    image_size = cv::Size(cal["image_width"].as<int>(), cal["image_height"].as<int>());
  }

  return CalibrationMatrices{camera_matrix, dist_coeffs, image_size};
}

fs::path saveCalibrationMatrices(const cv::Mat& camera_matrix,
                                 const cv::Mat& dist_coeffs,
                                 double reprojection_error,
                                 cv::Size image_size,
                                 const std::optional<fs::path>& config_path) {
  const fs::path path = resolveConfigPath(config_path);
  YAML::Node config = loadConfig(path);
  if (!config["calibration"]) {
    config["calibration"] = YAML::Node(YAML::NodeType::Map);
  }

  YAML::Node matrix_node(YAML::NodeType::Sequence);
  for (int r = 0; r < 3; ++r) {
    YAML::Node row(YAML::NodeType::Sequence);
    for (int c = 0; c < 3; ++c) {
      row.push_back(camera_matrix.at<double>(r, c));
    }
    matrix_node.push_back(row);
  }

  YAML::Node dist_node(YAML::NodeType::Sequence);
  cv::Mat flat = dist_coeffs.reshape(1, static_cast<int>(dist_coeffs.total()));
  for (int i = 0; i < flat.rows; ++i) {
    dist_node.push_back(flat.at<double>(i, 0));
  }

  config["calibration"]["enabled"] = true;
  config["calibration"]["camera_matrix"] = matrix_node;
  config["calibration"]["dist_coeffs"] = dist_node;
  config["calibration"]["reprojection_error"] = reprojection_error;
  config["calibration"]["image_width"] = image_size.width;
  config["calibration"]["image_height"] = image_size.height;

  return saveConfig(config, path);
}

}  // namespace chess_robot
